unit/solvers/strings/string_format_builtin_function/length_for_format_specifier.cpp

```diff
@@ -0,0 +1,260 @@
+#include <util/string_expr.h>
+/*******************************************************************\
+
+Module: Unit tests for length_for_format_specifier
+        solvers/refinement/string_format_builtin_function.cpp
+
+Author: Diffblue Ltd.
+
+\*******************************************************************/
+
+#include <testing-utils/use_catch.h>
+
+#include <solvers/strings/format_specifier.h>
+#include <solvers/strings/string_format_builtin_function.h>
+#include <util/simplify_expr.h>
+#include <util/symbol_table.h>
+
+// Create array_string_exprt from array of characters
+array_string_exprt from_char_vector(
+  std::vector<mp_integer> string_array,
+  const typet &char_type,
+  const typet &length_type,
+  const pointer_typet &pointer_type,
+  array_poolt &array_pool)
+{
+  const array_typet array_type(
+    char_type, from_integer(string_array.size(), length_type));
+  std::vector<exprt> expr_array;
+  expr_array.reserve(string_array.size());
+  std::transform(
+    string_array.begin(),
+    string_array.end(),
+    std::back_inserter(expr_array),
+    [&](const mp_integer el) { return from_integer(el, char_type); });
+  const array_exprt array(expr_array, array_type);
+  const exprt first_element = index_exprt(array, from_integer(0, length_type));
+  const exprt pointer = address_of_exprt(first_element, pointer_type);
+  const array_string_exprt arg = array_pool.find(pointer, array_type.size());
+  array_pool.insert(pointer, arg); // really necessary?
+  return arg;
+}
+
+SCENARIO(
+  "length_for_format_specifier",
+  "[core][solvers][strings][string_format_builtin_function]")
+{
+  const typet char_type = unsignedbv_typet(16);
+  const typet length_type = signedbv_typet(32);
+  const std::size_t pointer_width = 16;
+  const auto pointer_type = pointer_typet(char_type, pointer_width);
+
+  const symbol_tablet symbol_table;
+  const namespacet ns{symbol_table};
+  symbol_generatort fresh_symbol;
+  array_poolt array_pool{fresh_symbol};
+
+  WHEN("format specifier = %s")
+  {
+    WHEN("arg = serialized(\"bar\")")
+    {
+      array_string_exprt arg = from_char_vector(
+        {'b', 'a', 'r'}, char_type, length_type, pointer_type, array_pool);
+      format_specifiert fs{0, "", 0, 0, false, format_specifiert::STRING};
+      const exprt actual =
+        length_for_format_specifier(fs, arg, length_type, array_pool);
+      THEN("length constraint represents 3")
+      {
+        int to_int = *numeric_cast<int>(to_constant_expr(actual));
+        REQUIRE(to_int == 3);
+      }
+    }
+  }
+
+  WHEN("format specifier = %d")
+  {
+    WHEN("arg = serialized(12)")
+    {
+      const array_string_exprt arg = from_char_vector(
+        {0, 0, 0, 12}, char_type, length_type, pointer_type, array_pool);
+      format_specifiert fs{
+        0, "", 0, 0, false, format_specifiert::DECIMAL_INTEGER};
```

Comment:
:bulb: maybe defining a constructor for `format_specifiert` isn't so useful and writting `format_specifiert fs; fs.conversion = format_specifiert::DECIMAL_INTEGER;` would have been clearer.